tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +244,205 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = []
+    # We care about the order of dependencies list, so it will be a list.
+    collapsed_deps = set()
+    # We don't care about the order of srcs and headers, they will be sorted.
+    collapsed_srcs = set(_extract_sources(bazel_rule))
+    collapsed_public_headers = set(_extract_public_headers(bazel_rule))
+    collapsed_headers = set(_extract_nonpublic_headers(bazel_rule))
+    exclude_deps = set()
+
+    for dep in direct_deps:
+        external_dep_name_maybe = _external_dep_name_from_bazel_dependency(dep)
+
+        if dep in bazel_rules:
+            # Suppress expanding absl libs; expanding absl deps are technically
+            # correct but creates too much noise
+            if external_dep_name_maybe is None or not external_dep_name_maybe.startswith(
+                    'absl'):
+                if "_PROCESSING_DONE" not in bazel_rules[dep]:
+                    # This item is not processed before, compute now
+                    _compute_transitive_metadata(dep, bazel_rules,
+                                                 bazel_label_to_dep_name)
+                _deduplicate_append(
+                    transitive_deps,
+                    bazel_rules[dep].get('_TRANSITIVE_DEPS', []))
+                collapsed_deps.update(
+                    collapsed_deps, bazel_rules[dep].get('_COLLAPSED_DEPS', []))
+                exclude_deps.update(bazel_rules[dep].get('_EXCLUDE_DEPS', []))
+
+        # This dep is a public target, add it as a dependency
+        if dep in bazel_label_to_dep_name:
+            _deduplicate_append(transitive_deps, [bazel_label_to_dep_name[dep]])
+            collapsed_deps.update(collapsed_deps,
+                                  [bazel_label_to_dep_name[dep]])
+            # Add all the transitive deps of our every public dep to exclude
+            # list since we want to avoid building sources that are already
+            # built our dependencies
+            exclude_deps.update(bazel_rules[dep]['_TRANSITIVE_DEPS'])
+            exclude_deps.add(dep)
+            continue
 
-    target_name = target_dict['name']
-    bazel_deps = target_dict['_DEPS_BAZEL']
+        # This dep is an external target, add it as a dependency
+        if external_dep_name_maybe is not None:
+            _deduplicate_append(transitive_deps, [external_dep_name_maybe])
+            collapsed_deps.update(collapsed_deps, [external_dep_name_maybe])
+            continue
 
-    # initial values
-    public_headers = set(target_dict['_PUBLIC_HEADERS_BAZEL'])
-    headers = set(target_dict['_HEADERS_BAZEL'])
-    src = set(target_dict['_SRC_BAZEL'])
-    deps = set()
+    # Direct dependencies are part of transitive dependencies
+    _deduplicate_append(transitive_deps, direct_deps)
+
+    # Remove intermediate targets that our public dependencies already depend
```

Comment:
not sure I follow:  you're saying "remove intermediate targets", but from line 336  (`exclude_deps.add(dep)`) it seems that exclude_deps also contain public deps. So the comment seems inaccurate.

Q: if collapsed_deps contain public deps (which I think they need to), how come skipping the exclude deps here doesn't result in skipping our direct public dependency that really should have been included?

Example: 
 PA -> [PB, IX, IY]
on line 336, you add PB to exclude_deps and  then on this line, you will filter out PB from collapsed_deps.  So PA's deps will end up being empty, which is wrong collapsed deps should contain the info that PA -> PB.

Reply:
As I mentioned in https://github.com/grpc/grpc/pull/25758, the direct public dependency will be included in `exclude_dep`. It saves cycles to compute this list with recursion, but if you insist we can compute it for every node like https://github.com/grpc/grpc/pull/25758, which does improve the readability.